### PR TITLE
DDCNLS-6200: Migration to new Auth

### PR DESCRIPTION
**Pull request by nathan-maybrey**

### CODE REVIEW

File: test/controllers/auth/AuthActionSpec.scala [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+      val authAction = new AuthActionImpl(mockAuthConnector, app.configuration)
+      val controller = new Harness(authAction)
+      val result = controller.onPageLoad()(FakeRequest("", ""))
+      status(result) shouldBe SEE_OTHER
+      redirectLocation(result).get should endWith(ggSignInUrl)
+    }
+  }
+
+  "A user with confidence level below 50" should {
+    "return 303 and be redirected to GG sign in page" in {
```
Comment (daniel-manning) [inline test/controllers/auth/AuthActionSpec.scala]: I don't think this is a thing

File: test/controllers/auth/AuthActionSpec.scala [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+    }
+  }
+
+  "A user with a confidence level 50 and an SA enrolment" should {
+    "create an authenticated request" in {
+      val utr = new SaUtrGenerator().nextSaUtr.utr
+      val retrievalResult: Future[
+        Enrolments ~ Option[String] ~ Option[String]] =
+        Future.successful(
+            new ~(
```
Comment (daniel-manning) [inline test/controllers/auth/AuthActionSpec.scala]: you can bring in @jaynabonne RetrivalsOps to write this more cleanly

File: test/controllers/auth/MinAuthActionSpec.scala [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+      val minAuthAction = new MinAuthActionImpl(mockAuthConnector, app.configuration)
+      val controller = new Harness(minAuthAction)
+      val result = controller.onPageLoad()(FakeRequest("", ""))
+      status(result) shouldBe SEE_OTHER
+      redirectLocation(result).get should endWith(ggSignInUrl)
+    }
+  }
+
+  "A user with confidence level below 50" should {
+    "return 303 and be redirected to GG sign in page" in {
```
Comment (daniel-manning) [inline test/controllers/auth/MinAuthActionSpec.scala]: don't think this is a thing

File: test/services/AtsListServiceSpec.scala [hunk 4]
```diff
@@ -99,25 +94,23 @@ class AtsListServiceTest extends UnitSpec with FakeTaxsPlayApplication with Mock
...
 
-      when(dataCache.storeAtsTaxYearForSession(eqTo(2014))(any[HeaderCarrier], any[ExecutionContext]))
-        .thenReturn(Future.successful(None))
+      when(dataCache.storeAtsTaxYearForSession(eqTo(2014))(any[HeaderCarrier], any[ExecutionContext])).thenReturn(Future.successful(None))
 
       val result = storeSelectedTaxYear(2014)
 
       whenReady(result.failed) { exception =>
-        exception shouldBe a[NoSuchElementException]
+        exception shouldBe an [NoSuchElementException]
```
Comment (daniel-manning) [inline test/services/AtsListServiceSpec.scala]: a NoSuchElementException is right

File: test/services/AtsListServiceSpec.scala [hunk 5]
```diff
@@ -133,21 +126,19 @@ class AtsListServiceTest extends UnitSpec with FakeTaxsPlayApplication with Mock
 
     "Return a failed future when None is returned from the dataCache" in new TestService {
 
-      when(dataCache.fetchAndGetAtsTaxYearForSession(any[HeaderCarrier], any[ExecutionContext]))
-        .thenReturn(Future.successful(None))
+      when(dataCache.fetchAndGetAtsTaxYearForSession(any[HeaderCarrier], any[ExecutionContext])).thenReturn(Future.successful(None))
 
       whenReady(fetchSelectedTaxYear.failed) { exception =>
-        exception shouldBe a[NoSuchElementException]
+        exception shouldBe an [NoSuchElementException]
```
Comment (daniel-manning) [inline test/services/AtsListServiceSpec.scala]: ditto

File: app/controllers/auth/AuthAction.scala [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+          val vrn: Option[Vrn] = enrolments.find(vatEnrolments => vatEnrolments.key == "HMCE-VATDEC-ORG" || vatEnrolments.key == "HMCE-VATVAR-ORG")
+            .flatMap { enrolment =>
+              enrolment.identifiers
+                .find(id => id.key == "VATRegNo")
+                .map(key => Vrn(key.value))
+            }
+
+          block {
+            AuthenticatedRequest(
+              externalId,
```
Comment (daniel-manning) [inline app/controllers/auth/AuthAction.scala]: field in the analytics is userId, why are we using the externalId? Might be worth looking at the splunk with the analysts to see what that field looks like

File: app/controllers/IndexController.scala [hunk 7]
```diff
@@ -66,81 +70,71 @@ trait IndexController extends TaxsController {
         agentToken.fold[Future[_]] {
           Future.successful(None)
         } { token =>
-          AccountUtils.isAgent(user) match {
+          AccountUtils.isAgent(request) match {
             case true =>
               dataCache.storeAgentToken(token) recover { case e: Throwable => throw e }
             case _ =>
               Future.successful(None)
           }
```
[the remark this answers is not in the document]
Comment (nathan-maybrey) [reply app/controllers/IndexController.scala]: Done

File: app/controllers/auth/AuthAction.scala [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+          "continue"    -> Seq(callbackUrl),
+          "origin"      -> Seq(ApplicationConfig.appName)
+        )
+      )
+    }
+    case _: InsufficientConfidenceLevel => {
+      lazy val ggSignIn = ApplicationConfig.loginUrl
+      lazy val callbackUrl = ApplicationConfig.loginCallback
+      Redirect(
+        ggSignIn,
```
[the remark this answers is not in the document]
Comment (nathan-maybrey) [reply app/controllers/auth/AuthAction.scala]: Can we just remove this as confidence level is only required to be 50?

File: app/services/AtsService.scala [hunk 9]
```diff
@@ -45,52 +44,49 @@ trait AtsService {
...
+          if (accountUtils.isAgent(request)) {
+            dataCache.getAgentToken.flatMap {
+            token => getAtsAndStore(taxYear,token)
           }
+        } else {
+          getAtsAndStore(taxYear)
+        }
       }
     }
   } flatMap { identity }
```
[the remark this answers is not in the document]
Comment (nathan-maybrey) [reply app/services/AtsService.scala]: Done

File: app/services/AtsService.scala [hunk 10]
```diff
@@ -100,19 +96,17 @@ trait AtsService {
         getAtsAndStore(taxYear, token)
       }
     }
-  } flatMap (identity)
+  } flatMap(identity)
+
```
[the remark this answers is not in the document]
Comment (nathan-maybrey) [reply app/services/AtsService.scala]: Done

File: app/services/AtsService.scala [hunk 11]
```diff
@@ -100,19 +96,17 @@ trait AtsService {
...
+  private def getAtsAndStore(taxYear: Int, agentToken: Option[AgentToken] = None)(implicit hc: HeaderCarrier, request: AuthenticatedRequest[_]): Future[AtsData] = {
+    val account = utils.AccountUtils.getAccount(request)
     val requestedUTR = authUtils.getRequestedUtr(account, agentToken)
 
     //This warning is unchecked because we know that AuthorisedFor will only give us those accounts
     val gotData = (account: @unchecked) match {
-      case agent: TaxSummariesAgentAccount => middleConnector.connectToAtsOnBehalfOf(agent.uar, requestedUTR, taxYear)
-      case individual: SaAccount           => middleConnector.connectToAts(individual.utr, taxYear)
+      case agentUar: Uar => middleConnector.connectToAtsOnBehalfOf(agentUar, requestedUTR, taxYear)
+      case individualUtr: SaUtr => middleConnector.connectToAts(individualUtr, taxYear)
```
[the remark this answers is not in the document]
Comment (nathan-maybrey) [reply app/services/AtsService.scala]: yes we've checked and this works and passes all acceptance tests. We create the SaUtr and Uar in the AuthAction so they are safe to compare on type

File: app/services/AtsService.scala [hunk 12]
```diff
@@ -128,36 +122,30 @@ trait AtsService {
     }
   } flatMap { identity }
```
[the remark this answers is not in the document]
Comment (nathan-maybrey) [reply app/services/AtsService.scala]: Done

File: app/services/AtsListService.scala [hunk 13]
```diff
@@ -103,16 +101,13 @@ trait AtsListService {
...
-    val account = utils.AccountUtils.getAccount(user)
+  private def getAtsListAndStore(agentToken: Option[AgentToken]=None)(implicit hc: HeaderCarrier, request: AuthenticatedRequest[_]): Future[AtsListData] = {
+    val account = utils.AccountUtils.getAccount(request)
     val requestedUTR = authUtils.getRequestedUtr(account, agentToken)
 
     val gotData = (account: @unchecked) match {
-      case agent: TaxSummariesAgentAccount => middleConnector.connectToAtsListOnBehalfOf(agent.uar, requestedUTR)
-      case individual: SaAccount           => middleConnector.connectToAtsList(individual.utr)
+      case agent: Uar => middleConnector.connectToAtsListOnBehalfOf(agent, requestedUTR)
+      case individual: SaUtr => middleConnector.connectToAtsList(individual)
```
[the remark this answers is not in the document]
Comment (nathan-maybrey) [reply app/services/AtsListService.scala]: Ditto

File: app/services/TotalIncomeTaxService.scala [hunk 14]
```diff
@@ -16,64 +16,62 @@
...
-import uk.gov.hmrc.play.frontend.auth.{AuthContext => User}
 import utils.GenericViewModel
 import view_models.TotalIncomeTax
 
 import scala.concurrent.Future
 
 object TotalIncomeTaxService extends TotalIncomeTaxService {
-  override val atsService = AtsService
-  override val atsYearListService = AtsYearListService
+  override val atsService: AtsService.type = AtsService
```
[the remark this answers is not in the document]
Comment (nathan-maybrey) [reply app/services/TotalIncomeTaxService.scala]: Do you mean the type annotation? if so it's done now

File: test/controllers/auth/AuthActionSpec.scala (hunk 1 above)
Comment (nathan-maybrey) [reply test/controllers/auth/AuthActionSpec.scala]: removed

File: test/controllers/auth/AuthActionSpec.scala (hunk 2 above)
Comment (nathan-maybrey) [reply test/controllers/auth/AuthActionSpec.scala]: Done

File: test/controllers/auth/MinAuthActionSpec.scala (hunk 3 above)
Comment (nathan-maybrey) [reply test/controllers/auth/MinAuthActionSpec.scala]: removed

File: test/services/AtsListServiceSpec.scala (hunk 4 above)
Comment (nathan-maybrey) [reply test/services/AtsListServiceSpec.scala]: Done

File: test/services/AtsListServiceSpec.scala (hunk 5 above)
Comment (nathan-maybrey) [reply test/services/AtsListServiceSpec.scala]: done

File: app/controllers/auth/AuthAction.scala (hunk 6 above)
Comment (nathan-maybrey) [reply app/controllers/auth/AuthAction.scala]: No UserId available from retrievals, Confluence guide said to use either InternalId or ExternalId to replace UserId, both of which are just a random number.

Comment (daniel-manning) [reply app/controllers/auth/AuthAction.scala]: Ok, great

File: app/controllers/auth/AuthAction.scala (hunk 8 above)
[the remark this answers is not in the document]
Comment (daniel-manning) [reply app/controllers/auth/AuthAction.scala]: yes, makes sense

File: app/controllers/auth/MinAuthAction.scala [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+      lazy val callbackUrl = ApplicationConfig.loginCallback
+      Redirect(
+        ggSignIn,
+        Map(
+          "continue" -> Seq(callbackUrl),
+          "origin" -> Seq(ApplicationConfig.appName)
+        )
+      )
+    }
+    case _: InsufficientConfidenceLevel => {
```
[the remark this answers is not in the document]
Comment (daniel-manning) [reply app/controllers/auth/MinAuthAction.scala]: removed in my commit

**Pull request merged**